// Copyright 2023 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package kuscia.proto.api.v1alpha1.kusciaapi;

import "kuscia/proto/api/v1alpha1/common.proto";

option go_package = "github.com/secretflow/kuscia/proto/api/v1alpha1/kusciaapi";
option java_package = "org.secretflow.v1alpha1.kusciaapi";

service DomainService {
  rpc CreateDomain(CreateDomainRequest) returns (CreateDomainResponse);

  rpc QueryDomain(QueryDomainRequest) returns (QueryDomainResponse);

  rpc UpdateDomain(UpdateDomainRequest) returns (UpdateDomainResponse);

  rpc DeleteDomain(DeleteDomainRequest) returns (DeleteDomainResponse);

  rpc BatchQueryDomain(BatchQueryDomainRequest) returns (BatchQueryDomainResponse);
}

message CreateDomainRequest {
  RequestHeader header = 1;
  string domain_id = 2;
  string role = 3;
  string cert = 4;
  AuthCenter auth_center = 5;
  string master_domain_id = 6;
}

message CreateDomainResponse {
  Status status = 1;
}

message DeleteDomainRequest {
  RequestHeader header = 1;
  string domain_id = 2;
}

message DeleteDomainResponse {
  Status status = 1;
}

message QueryDomainRequest {
  RequestHeader header = 1;
  string domain_id = 2;
}

message QueryDomainResponse {
  Status status = 1;
  QueryDomainResponseData data = 2;
}

message QueryDomainResponseData {
  string domain_id = 1;
  string role = 2;
  string cert = 3;
  repeated NodeStatus node_statuses = 4;
  repeated DeployTokenStatus deploy_token_statuses = 5;
  map<string, string> annotations = 6;
  AuthCenter auth_center = 7;
  string master_domain_id = 8;
}

message NodeStatus {
  string name = 1;
  string status = 2;
  string version = 3;
  string last_heartbeat_time = 4;
  string last_transition_time = 5;
}

message UpdateDomainRequest {
  RequestHeader header = 1;
  string domain_id = 2;
  string role = 3;
  string cert = 4;
  AuthCenter auth_center = 5;
  string master_domain_id = 6;
}

message UpdateDomainResponse {
  Status status = 1;
}

message BatchQueryDomainRequest {
  RequestHeader header = 1;
  repeated string domain_ids = 2;
}

message BatchQueryDomainResponse {
  Status status = 1;
  BatchQueryDomainResponseData data = 2;
}

message BatchQueryDomainResponseData {
  repeated Domain domains = 1;
}

message Domain {
  string domain_id = 1;
  string role = 2;
  string cert = 3;
  repeated NodeStatus node_statuses = 4;
}

message DeployTokenStatus {
  // deploy token
  string token = 1;
  // used or unused
  string state = 2;
  string last_transition_time = 3;
}

message AuthCenter {
  // Token
  string authentication_type = 1;
  // UID-RSA-GEN
  string token_gen_method = 2;
}